File: CS2040/factorfree.java
import java.util.*;
import java.io.*;

public class factorfree {
    public static void main(String[] args) throws IOException {
        var br = new BufferedReader(new InputStreamReader(System.in));
        var pw = new PrintWriter(System.out);

        int n = Integer.parseInt(br.readLine());
        var elements = new int[n + 1];
        var st = new StringTokenizer(br.readLine());
        int max = 0;
        for (int i = 1; i <= n; i++) {
            elements[i] = Integer.parseInt(st.nextToken());
            max = Math.max(max, elements[i]);
        }
        var factor = new int[max + 1];
        var primes = new ArrayList<Integer>();
        primes.add(0);
        sieve(factor, primes, max);

        int biggest = -1;
        for (int i : factor)
            biggest = Math.max(biggest, i);

        int[] pos = new int[biggest + 1], prev = new int[n + 1], next = new int[n + 1], parent = new int[n + 1];

        Arrays.fill(pos, 1, primes.size(), -1);
        for (int i = 1; i <= n; i++) {
            prev[i] = 1;
            for (int j = elements[i]; j > 1; j /= primes.get(factor[j]))
                prev[i] = Math.max(prev[i], pos[factor[j]] + 1);
            for (int j = elements[i]; j > 1; j /= primes.get(factor[j]))
                pos[factor[j]] = i;
        }
        Arrays.fill(pos, 1, primes.size(), n + 1);
        for (int i = n; i > 0; i--) {
            next[i] = n;
            for (int j = elements[i]; j > 1; j /= primes.get(factor[j]))
                next[i] = Math.min(next[i], pos[factor[j]] - 1);
            for (int j = elements[i]; j > 1; j /= primes.get(factor[j]))
                pos[factor[j]] = i;
        }
        if (dfs(1, n, 0, prev, next, parent)) {
            for (int i = 1; i <= n; i++) pw.print(parent[i] + " ");
            pw.println();
        } else pw.println("impossible");
        pw.flush();
    }

    static void sieve(int[] factor, ArrayList<Integer> primes, int max_element) {
        var prime = new BitSet(max_element + 1);
        prime.set(2, max_element + 1);
        for (int i = 2; i <= max_element; i++) {
            if (prime.get(i)) {
                primes.add(i);
                factor[i] = primes.size() - 1;
            }
            for (int j = 1, k; j < primes.size() && (k = i * primes.get(j)) <= max_element; j++) {
                factor[k] = factor[i];
                prime.clear(k);
                if (i % primes.get(j) == 0) break;
            }
        }
    }

    static boolean dfs(int l, int r, int p, int[] prev, int[] next, int[] parent) {
        var s = new Stack<Integer>();
        s.push(l);
        s.push(r);
        s.push(p);

        while (!s.isEmpty()) {
            p = s.pop();
            r = s.pop();
            l = s.pop();

            if (l > r) continue;
            
            var found = false;
            for (int left = l, right = r; left <= right; left++, right--) 
                if ((l >= prev[left] && r <= next[left]) || (l >= prev[right] && r <= next[right])) {
                    int mid = (l >= prev[left] && r <= next[left]) ? left : right;
                    parent[mid] = p;

                    s.push(mid + 1);
                    s.push(r);
                    s.push(mid);

                    s.push(l);
                    s.push(mid - 1);
                    s.push(mid);

                    found = true;
                    break;
                }
            
            if (!found) return false;
        }

        return true;
    }
}

File: CS3233/factorfree.cpp
#include <bits/stdc++.h>
using namespace std;

bool dfs(int l, int r, vector<int> &prev, vector<int> &left, vector<int> &right, int p = -1) {
    if (l + 1 > r) return true;
    
    for (int i = l, j = r - 1; i <= j; i++, j--) {
        prev[i] = p;
        if (left[i] < l && right[i] >= r) return dfs(l, i, prev, left, right, i) && dfs(i + 1, r, prev, left, right, i);
        
        prev[j] = p;
        if (left[j] < l && right[j] >= r) return dfs(l, j, prev, left, right, j) && dfs(j + 1, r, prev, left, right, j);
    }

    return false;
}

vector<int> sieve(int n) {
    vector<int> lpf(n + 1, 0);
    for (int i = 2; i <= n; i++)
        if (!lpf[i])
            for (int j = i; j <= n; j += i) lpf[j] = i;

    return lpf;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    vector<int> elements(n);
    for (int &a : elements) cin >> a;

    int biggest = *max_element(elements.begin(), elements.end());
    vector<int> lpf = sieve(biggest), prev(biggest + 1, -1), left(n, -1), right(n, n), prev(n, 0);
    for (int i = 0; i < n; i++)
        for (int num = elements[i]; num > 1; num /= lpf[num]) {
            if (prev[lpf[num]] < i) {
                left[i] = max(left[i], prev[lpf[num]]);
                if (prev[lpf[num]] != -1) right[prev[lpf[num]]] = min(right[prev[lpf[num]]], i);
            }
            prev[lpf[num]] = i;
        }

    if (dfs(0, n, prev, left, right))
        for (int p : prev) cout << p + 1 << " ";
    else cout << "impossible";
}
